File: processing/syntactic/expressions/atomic/name_expr.py
from __future__ import annotations

from typing import List, Union

from processing.codegen.code_builder import CodeBuilder
from processing.lexical.tokens.token import Token
from processing.lexical.tokens.token_type import TokenType
from processing.location import span_marker
from processing.syntactic.expressions.expr import Expr
from processing.syntactic.expressions.groups import VarTargetExpression


class NameExpr(VarTargetExpression):
    """ name_expr:
        ID {'.' ID};
    """

    @property
    def is_simple(self) -> bool:
        return len(self.qualifiers) > 0

    @property
    def qualifiers(self) -> List[Token]:
        return [t for t in self.name_parts if t.ttype == TokenType.identifier]

    @property
    def value_type(self):
        from processing.syntactic.expressions.block_expr import BlockExpr

        x = self.get_parent_of_type(BlockExpr).get_def_by_name(self)
        if x is not None:
            return x.value_type

    def __init__(
            self,
            parent: Expr = None,
            name: Union[List[Token], str] = None
    ):
        super().__init__(parent)
        if isinstance(name, str):
            self.name_parts = []
            qs = name.split('.')
            for i, q in enumerate(qs):
                self.name_parts.append(Token(self.source, TokenType.identifier, q))
                if i != len(qs) - 1:
                    self.name_parts.append(Token(self.source, TokenType.op_dot, '.'))
        else:
            self.name_parts = name or []

    def __str__(self):
        return '.'.join(t.value for t in self.qualifiers)

    def __repr__(self):
        return f"{self.__class__.__name__} ('{'.'.join(t.value for t in self.qualifiers)}')"

    @span_marker
    def parse(self, must_be_simple = False) -> NameExpr:
        self.stream.eat(TokenType.identifier)
        self.name_parts.append(self.stream.token)
        if not must_be_simple:
            while self.stream.maybe_eat(TokenType.op_dot):
                self.name_parts.append(self.stream.token)
                if self.stream.eat(TokenType.identifier):
                    self.name_parts.append(self.stream.token)
        return self

    def to_axion(self, c: CodeBuilder):
        c += self.name_parts

    def to_csharp(self, c: CodeBuilder):
        if self.is_simple and self.name_parts[0].value == 'self':
            c += 'this'
            return
        c += self.name_parts

    def to_python(self, c: CodeBuilder):
        c += self.name_parts

File: Axion.Core/Processing/Syntactic/Expressions/Atomic/NameExpr.cs
using System.Linq;
using Axion.Core.Processing.Lexical.Tokens;
using Axion.Core.Processing.Syntactic.Expressions.Common;
using Axion.Core.Processing.Syntactic.Expressions.TypeNames;
using Magnolia.Attributes;
using Magnolia.Trees;
using static Axion.Specification.TokenType;

namespace Axion.Core.Processing.Syntactic.Expressions.Atomic;

/// <summary>
///     <code>
///         name-expr:
///             ID {'.' ID};
///     </code>
/// </summary>
[Branch]
public partial class NameExpr : AtomExpr {
    [Leaf] NodeList<Token, Ast>? tokens;

    public Token[] Qualifiers => Tokens.OfType(Identifier);
    public bool IsSimple => Qualifiers.Length == 1;

    public override TypeName? InferredType =>
        ((Node?) GetParent<ScopeExpr>()?.GetDefByName(this))?.InferredType;

    public NameExpr(Node parent) : base(parent) { }

    public NameExpr(string name) : this(null, name) { }

    public NameExpr(Node? parent, string name) : base(parent) {
        if (name.Contains('.')) {
            var qs = name.Split('.');
            for (var i = 0; i < qs.Length; i++) {
                var q = qs[i];
                Tokens += new Token(Unit, Identifier, q);
                if (i != qs.Length - 1) {
                    Tokens += new Token(Unit, Dot, ".");
                }
            }
        }
        else {
            Tokens += new Token(Unit, Identifier, name);
        }
    }

    public static implicit operator NameExpr(string name) {
        return new(null, name);
    }

    public NameExpr Parse(bool simple = false) {
        Tokens += Stream.Eat(Identifier);
        if (simple) {
            return this;
        }

        while (Stream.PeekIs(Dot)) {
            Tokens += Stream.Eat();
            Tokens += Stream.Eat(Identifier);
        }

        return this;
    }

    public override string ToString() {
        return string.Join(".", Qualifiers.Select(q => q.Content));
    }
}
